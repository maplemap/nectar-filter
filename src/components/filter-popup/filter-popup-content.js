import React, {useState, useCallback, useEffect} from 'react';
import {Button} from 'antd';

export const FilterPopupContent = ({
    filterType,
    filtersByType,
    appliedFiltersByType,
    popoverShown,
    onApply,
    onReset
}) => {
    const [checkedFiltersIds, setCheckedFiltersIds] = useState([]);

    useEffect(() => {
        if (!popoverShown) {
            setCheckedFiltersIds(appliedFiltersByType || []);
        }
    }, [popoverShown, appliedFiltersByType]);

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const itemOnClick = useCallback(e => {
        const {name} = e.target;
        const newCheckedFilters = (
            checkedFiltersIds.includes(name) ?
                checkedFiltersIds.filter(filter => filter !== name) :
                [...checkedFiltersIds, name]
        );

        setCheckedFiltersIds(newCheckedFilters);
    });

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const setAppliedFilters = useCallback(() => {
        onApply({checkedFiltersIds, filterType});
    });

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const resetCheckedFilters = useCallback(() => {
        setCheckedFiltersIds([]);
        onReset(filterType);
    });

    return (
        <>
            <div>
                {filtersByType.map(({id, title}) => (
                    <Button key={id} {...{
                        name: id,
                        className: 'filter-popup__content-item',
                        onClick: itemOnClick,
                        ...(checkedFiltersIds.includes(id) ? {type: 'primary'} : null)
                    }}
                    >
                        {title}
                    </Button>
                ))}
            </div>
            <div className="filter-popup__action-buttons">
                {checkedFiltersIds.length > 0 &&
                    <Button onClick={resetCheckedFilters}>Cancel</Button>}
                <Button type="primary" onClick={setAppliedFilters}>Apply</Button>
            </div>
        </>
    );
};
