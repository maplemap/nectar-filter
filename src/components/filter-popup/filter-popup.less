@import '../../app/mixins';

.filter-popup {

  &__content-item {
     margin: 0 .6rem .6rem 0;
  }

  &__action-buttons {
    display: flex;
    justify-content: space-between;
    margin-top: 1rem;
    padding-top: 1rem;
    border-top: @color-gray 1px solid;

    button:last-child {
      margin-left: auto;
    }
  }
}
