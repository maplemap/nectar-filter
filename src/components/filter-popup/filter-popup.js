import React, {useContext, useState, useCallback, useEffect} from 'react';
import {Popover, Button} from 'antd';
import {ContextApp} from '../../reducer';

import './filter-popup.less';

const PRIMARY = 'primary';

export const FilterPopup = ({children, filterType, ...props}) => {
    const [checkedFiltersIds, setCheckedFiltersIds] = useState([]);
    const {state: {filtersData, appliedFilters}, dispatch} = useContext(ContextApp);

    useEffect(() => {
        if (!appliedFilters[filterType]) {
            setCheckedFiltersIds([]);
        }
    }, [appliedFilters]);

    const itemOnClick = useCallback(e => {
        const {name} = e.target;
        const newCheckedFilters = (
            checkedFiltersIds.includes(name) ?
                checkedFiltersIds.filter(filter => filter !== name) :
                [...checkedFiltersIds, name]
        );

        setCheckedFiltersIds(newCheckedFilters);
    });

    const setAppliedFilters = useCallback(() => {
        if (checkedFiltersIds.length > 0) {
            dispatch({
                appliedFilters: {
                    ...appliedFilters,
                    [filterType]: checkedFiltersIds
                }
            });
        }
    });

    const resetCheckedFilters = useCallback(() => {
        const newAppliedFilters = {...appliedFilters};
        delete newAppliedFilters[filterType];

        dispatch({appliedFilters: newAppliedFilters});
        setCheckedFiltersIds([]);
    });

    const onVisibleChange = shown => {
        if (!shown) {
            setCheckedFiltersIds(appliedFilters[filterType] || []);
        }

        props.onVisibleChange(shown);
    };

    const getContent = () => (
        <>
            <div>
                {filtersData[filterType].map(({id, title}) => (
                    <Button key={id} {...{
                        name: id,
                        className: 'filter-popup__content-item',
                        onClick: itemOnClick,
                        ...(checkedFiltersIds.includes(id) ? {type: PRIMARY} : null)
                    }}
                    >
                        {title}
                    </Button>
                ))}
            </div>
            <div className="filter-popup__action-buttons">
                {
                    checkedFiltersIds.length > 0 &&
                    <Button onClick={resetCheckedFilters}>Cancel</Button>
                }
                <Button type="primary" onClick={setAppliedFilters}>
                    Apply
                </Button>
            </div>
        </>
    );

    return (
        <Popover
            placement="bottom"
            trigger="click"
            content={getContent()}
            onVisibleChange={onVisibleChange}
            overlayStyle={{width: '20rem', padding: 0}}
        >
            {children}
        </Popover>
    );
};
