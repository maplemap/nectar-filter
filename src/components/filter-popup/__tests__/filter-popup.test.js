import React from 'react';
import {shallow} from 'enzyme';
import {Popover} from 'antd';
import {FilterPopup} from '../filter-popup';

jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: jest.fn().mockImplementation(() => ({state: {filtersData: {}, appliedFilters: {}}})),
    useState: jest.fn()
}));

const getComponent = props => (
    <FilterPopup {...{
        ...props
    }}
    />
);
const getShallow = props => shallow(getComponent(props));

describe('FilterPopup', () => {
    let getContextData;
    let setPopoverShown;
    let dispatch;

    beforeAll(() => {
        dispatch = jest.fn();
        setPopoverShown = jest.fn();
        getContextData = ({filtersData = {}, appliedFilters = {}, isMobile = false}) => ({
            state: {filtersData, appliedFilters, isMobile},
            dispatch
        });

        React.useState.mockImplementation(popoverShown => [popoverShown, setPopoverShown]);
    });

    it('should render `Popover` component', () => {
        expect(getShallow().find(Popover).length).toBe(1);
    });

    describe('given the `filterTypes` is Array', () => {
        it('should render `Collapse` component as content of `Popover`', () => {
            expect(getShallow({filterTypes: ['type']}).find(Popover).props().content).toMatchSnapshot();
        });
    });

    describe('given the `filterTypes` is not Array', () => {
        it('should render `FilterPopupContent` component as content of `Popover`', () => {
            expect(getShallow({filterTypes: 'type'}).find(Popover).props().content).toMatchSnapshot();
        });
    });

    describe('when `onVisibleChange` method was invoked', () => {
        let onVisibleChange;
        let wrapper;
        const shown = true;

        beforeAll(() => {
            onVisibleChange = jest.fn();
            wrapper = getShallow({onVisibleChange});
            wrapper.find(Popover).props().onVisibleChange(shown);
        });

        it('should call `setPopoverShown` with `shown` parameter', () => {
            expect(setPopoverShown).toHaveBeenCalledWith(shown);
        });
        it('should call `onVisibleChange` property with `shown` parameter', () => {
            expect(onVisibleChange).toHaveBeenCalledWith(shown);
        });
    });

    describe('when `resetCheckedFilters` method was invoked', () => {
        it('should call `dispatch` with new `appliedFilters` without data by `filterType`', () => {
            const filterType = 'size';
            const appliedFilters = {color: {}, [filterType]: {}};
            React.useContext.mockReturnValueOnce(getContextData({appliedFilters}));

            const popoverContent = getShallow().find(Popover).props().content;
            popoverContent.props.onReset(filterType);
            delete appliedFilters[filterType];

            expect(dispatch).toHaveBeenCalledWith({appliedFilters});
            dispatch.mockReset();
        });
    });

    describe('when `setAppliedFilters` method was invoked', () => {
        describe('when `checkedFiltersIds` parameter has length more then `0`', () => {
            it('should call `dispatch` with new `appliedFilters` with data by `filterType`', () => {
                const filterType = 'size';
                const checkedFiltersIds = ['1x1'];
                const appliedFilters = {color: {}};
                React.useContext.mockReturnValueOnce(getContextData({appliedFilters}));

                const popoverContent = getShallow().find(Popover).props().content;
                popoverContent.props.onApply({checkedFiltersIds, filterType});

                expect(dispatch).toHaveBeenCalledWith({
                    appliedFilters: {...appliedFilters, ...{[filterType]: checkedFiltersIds}}
                });
                dispatch.mockReset();
            })
        });
    });
});
