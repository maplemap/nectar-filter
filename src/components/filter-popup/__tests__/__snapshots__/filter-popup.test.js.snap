// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterPopup given the \`filterTypes\` is Array should render \`Collapse\` component as content of \`Popover\` 1`] = `
<Collapse
  bordered={true}
  expandIconPosition="left"
  openAnimation={
    Object {
      "appear": [Function],
      "enter": [Function],
      "leave": [Function],
    }
  }
>
  <CollapsePanel
    header="Type"
  >
    <FilterPopupContent
      filterType="type"
      onApply={[Function]}
      onReset={[Function]}
      popoverShown={false}
    />
  </CollapsePanel>
</Collapse>
`;

exports[`FilterPopup given the \`filterTypes\` is not Array should render \`FilterPopupContent\` component as content of \`Popover\` 1`] = `
<FilterPopupContent
  filterType="type"
  onApply={[Function]}
  onReset={[Function]}
  popoverShown={false}
/>
`;
